### PR TITLE
HTTPCORE-761: support ExtendedSocketOption in socket 

### PR DESCRIPTION
**Pull request by kkewwei**
Support setting TCP_KEEPIDLE, TCP_KEEPINTERVAL, TCP_KEEPCOUNT in Socket.

### CODE REVIEW

**Comment by kkewwei** [comment]
@ok2c, please have a look when you are free.

**Comment by ok2c** [comment]
@kkewwei I am fine with the proposed changes. There is a problem with the build, however.

**Comment by ok2c** [comment]
@kkewwei One last thing. Could you please make symmetric changes to `SocketConfig` used by the classic transport?

**Comment by kkewwei** [comment]
@ok2c I have added the parameters in `SocketConfig`, but they are not used anywhere, is that ok?

**Comment by ok2c** [comment]
> @ok2c I have added the parameters in `SocketConfig`, but they are not used anywhere, is that ok?

@kkewwei Why? It is being used in `HttpRequester`, `HttpServer`  and `RequestListener`. Am I missing something?

**Comment by kkewwei** [comment]
> > @ok2c I have added the parameters in `SocketConfig`, but they are not used anywhere, is that ok?
> 
> @kkewwei Why? It is being used in `HttpRequester`, `HttpServer` and `RequestListener`. Am I missing something?

@ok2c, If the three parameters are set in `Socket`, I have to use `Socket.setOption()`, which is import in java9.
image

Is there any suggestion?

**Comment by ok2c** [comment]
> Is there any suggestion?

@kkewwei I suppose one has to use reflection like with `SocketOptions`.

File: httpcore5/src/main/java/org/apache/hc/core5/util/ReflectionUtils.java [hunk 1]
```diff
@@ -72,4 +75,35 @@ public static int determineJRELevel() {
...
+            return null;
+        }
+    }
+
+    /**
+     * object can be ServerSocket or Socket
+     *
+     * @since 5.3
+     */
+    public static <T> void setOption(final T object, final String fieldName, final T value) {
```
Comment (kkewwei) [inline httpcore5/src/main/java/org/apache/hc/core5/util/ReflectionUtils.java:96]: @ok2c whether  I should use `ReflectionUtils.callSetter` directly? 

But the method doesn't throw any exception, which is not suitable for me, I want an exception to be thrown if there is a problem.

I don't know whether I should change the logic of `ReflectionUtils.callSetter`: the method doesn't swallow the exception, which will have an impact on existing usage.

**Comment by kkewwei** [comment]
@ok2c please have a look when you are free.

File: httpcore5/src/main/java/org/apache/hc/core5/http/impl/bootstrap/HttpRequester.java [hunk 2]
```diff
@@ -82,9 +82,14 @@
 import org.apache.hc.core5.pool.PoolEntry;
 import org.apache.hc.core5.pool.PoolStats;
 import org.apache.hc.core5.util.Args;
+import org.apache.hc.core5.util.ReflectionUtils;
 import org.apache.hc.core5.util.TimeValue;
 import org.apache.hc.core5.util.Timeout;
 
+import static org.apache.hc.core5.reactor.SingleCoreIOReactor.TCP_KEEPCOUNT;
```
Comment (ok2c) [inline httpcore5/src/main/java/org/apache/hc/core5/http/impl/bootstrap/HttpRequester.java:89]: @kkewwei This should be avoided. Now classic transport services are directly coupled to a async transport class

File: httpcore5/src/main/java/org/apache/hc/core5/reactor/SingleCoreIOReactor.java [hunk 3]
```diff
@@ -51,8 +53,14 @@
 import org.apache.hc.core5.util.Args;
 import org.apache.hc.core5.util.Timeout;
 
-class SingleCoreIOReactor extends AbstractSingleCoreIOReactor implements ConnectionInitiator {
+import static org.apache.hc.core5.util.ReflectionUtils.getExtendedSocketOptionOrNull;
 
+@Internal
+public class SingleCoreIOReactor extends AbstractSingleCoreIOReactor implements ConnectionInitiator {
+
+    public static final String TCP_KEEPIDLE = "TCP_KEEPIDLE";
```
Comment (ok2c) [inline httpcore5/src/main/java/org/apache/hc/core5/reactor/SingleCoreIOReactor.java:61]: @kkewwei This is not the right place. Please move those constants to a separate file and mark it internal. The `org.apache.hc.core5.io` might be a good place.

File: httpcore5/src/main/java/org/apache/hc/core5/util/ReflectionUtils.java [hunk 4]
```diff
@@ -44,12 +46,13 @@ public static void callSetter(final Object object, final String setterName, fina
         }
     }
 
-    public static <T> T callGetter(final Object object, final String getterName, final Class<T> resultType) {
+    public static <T> T callGetter(final Object object, final String getterName, final Object arg, final Class argType, final Class<T> resultType) {
```
Comment (ok2c) [inline httpcore5/src/main/java/org/apache/hc/core5/util/ReflectionUtils.java:49]: @kkewwei Is this method actually being used anywhere besides tests? Do we need to change it?

File: httpcore5/src/main/java/org/apache/hc/core5/util/ReflectionUtils.java [hunk 5]
```diff
@@ -72,4 +75,35 @@ public static int determineJRELevel() {
...
     }
 
+    /**
+     * @since 5.3
+     */
+    public static <T> SocketOption<T> getExtendedSocketOptionOrNull(final String fieldName) {
+        try {
+            final Class<?> extendedSocketOptionsClass = Class.forName("jdk.net.ExtendedSocketOptions");
+            final Field field = extendedSocketOptionsClass.getField(fieldName);
+            return (SocketOption)field.get((Object)null);
```
Comment (ok2c) [inline httpcore5/src/main/java/org/apache/hc/core5/util/ReflectionUtils.java:85]: @kkewwei Please suppress `unchecked` warning

File: httpcore5/src/main/java/org/apache/hc/core5/util/ReflectionUtils.java [hunk 6]
```diff
@@ -72,4 +75,35 @@ public static int determineJRELevel() {
...
+     * object can be ServerSocket or Socket
+     *
+     * @since 5.3
+     */
+    public static <T> void setOption(final T object, final String fieldName, final T value) {
+        try {
+            final Class<?> serverSocketClass = object.getClass();
+            final Method setOptionMethod = serverSocketClass.getMethod("setOption", SocketOption.class, Object.class);
+            final SocketOption<Integer> tcpKeepIdle = getExtendedSocketOptionOrNull(fieldName);
+            if (tcpKeepIdle == null) {
```
Comment (ok2c) [inline httpcore5/src/main/java/org/apache/hc/core5/util/ReflectionUtils.java:101]: @kkewwei The choice of the name looks bad. Can you come up with a better name here?

File: httpcore5/src/main/java/org/apache/hc/core5/util/ReflectionUtils.java [hunk 7]
```diff
@@ -72,4 +75,35 @@ public static int determineJRELevel() {
...
+                throw new UnsupportedOperationException(fieldName + " is not supported in the current jdk");
+            }
+            setOptionMethod.invoke(object, tcpKeepIdle, value);
+        } catch (final Exception ignore) {
```
Comment (ok2c) [inline httpcore5/src/main/java/org/apache/hc/core5/util/ReflectionUtils.java:105]: @kkewwei `ignore` looks like a bad name here.

File: httpcore5/src/test/java/org/apache/hc/core5/reactor/TestSingleCoreIOReactor.java [hunk 8]
```diff
@@ -0,0 +1,68 @@
+/*
+ * ====================================================================
```
Comment (ok2c) [inline httpcore5/src/test/java/org/apache/hc/core5/reactor/TestSingleCoreIOReactor.java:2]: @kkewwei I am not sure what this test case actually tests. It does not seem to test much of `SingleCoreIOReactor`, just a single reflective call. Drop it.

File: httpcore5/src/test/java/org/apache/hc/core5/util/TestReflectionUtils.java [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+
+    @Test
+    public void testGetExtendedSocketOptionOrNull() {
+        testGetExtendedSocketOption(TCP_KEEPIDLE);
+        testGetExtendedSocketOption(TCP_KEEPINTERVAL);
+        testGetExtendedSocketOption(TCP_KEEPCOUNT);
+    }
+
+    private void testGetExtendedSocketOption(final String option) {
+        final SocketOption socketOption = getExtendedSocketOptionOrNull(option);
```
Comment (ok2c) [inline httpcore5/src/test/java/org/apache/hc/core5/util/TestReflectionUtils.java:57]: @kkewwei Please fix `SocketOption` generic declaration.

**Review by ok2c — CHANGES_REQUESTED** [review]
@kkewwei There is whole bunch of small things. Please take a look at your convenience.

File: httpcore5/src/main/java/org/apache/hc/core5/util/ReflectionUtils.java (hunk 4 above)
Comment (kkewwei) [reply httpcore5/src/main/java/org/apache/hc/core5/util/ReflectionUtils.java:49]: It seems unreasonable to change the publish method. otheres may be depend on it.

Should I create the similar method?

Comment (kkewwei) [reply httpcore5/src/main/java/org/apache/hc/core5/util/ReflectionUtils.java:49]: It indeed seem unreasonable to change the publish method. others may be depend on it.

**Comment by kkewwei** [comment]
@ok2c please have an another look.

**Comment by kkewwei** [comment]
@ok2c It's ok now, please have a another look.

**Review by ok2c — APPROVED** [review]

**Comment by ok2c** [comment]
@kkewwei I tweaked your code a little and moved new methods around

**Pull request merged**